### DISCUSSION TITLE
Using a same input parameter in different Material

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#23200] [Q&A General] [answered]
Hi all,

I'd like to use the same input parameters in two different source codes of Materials, but I don't want to define these parameters twice. I just want to define these in one of the source codes, the other material code can directly use them. For example, the parameters is shown as: 

**MooseEnum b_choice("1 2 3 ", "1");
params.addParam<MooseEnum>("b",  b_choice, " ");**

What should I do?

Thank you!

**Comment by GiudGiud** [comment, marked as answer]
Hello

you can create a base class that has these parameters. 
then either one of your material is the base class and the other is a derived class or both are derived classes. 

Guillaume

**Reply by SomnusYu** [reply]
Hi

Do you have an example of the base class material (AD type)? I have tried this way but the computeQpProperties function in the base class may cause the error. I don't know how to fix it.

Thank you!

**Reply by GiudGiud** [reply]
what error?
You do not necessarily need computeQpProperties defined in the base class as long as it s an abstract class (eg it s not meant to be instantiated in an input file, you dont register it to MooseApp). You will need computeQpProperties defined in all the classes that are registered

**Reply by SomnusYu** [reply]
But it is necessary to contain the computeQpProperties if we inherited the ADMaterial. Do you have an example?

**Reply by GiudGiud** [reply]
only if the class is registered. The base class could not define it.
If you want you can just define it as an empty routine, but make sure to override it in the derived classes

**Reply by SomnusYu** [reply]
Sorry, I'm not clear on what should I do. How to define it as an empty routine? If you kindly offer me an example, it would be clearer.

**Reply by GiudGiud** [reply]
`void <className>::computeQpProperties() {} `
for the empty one

then in the derived classes:
```
void <derivedClassName>::computeQpProperties()
{
  some_property[_qp] = value it needs;
}
```

**Reply by SomnusYu** [reply]
Oh, I see. Thank you very much!
